### ISSUE TITLE
Just log crash on iOS 12

### ISSUE THREAD

**Issue by phamsontm9x** [justeat/JustLog#110]
BUG IN CLIENT OF LIBDISPATCH: Unbalanced call to dispatch_group_leave()
![image](https://user-images.githubusercontent.com/11141953/125552896-5a11e35f-b692-4d59-b542-65ba4cc7e837.png)

**Comment by gigisommo** [comment]
@thanhsoncun Thank you for the crash report. We are aware of this crash. We just had few occurrences of this in production and we weren't able to reproduce it. Do you know how to reproduce it? How often does it happen?

**Comment by phamsontm9x** [comment]
@gigisommo Thank you for your support. My users had crashed many times when using app. I have problem when i updated justlog 3.4 -> 3.6.

**Comment by gigisommo** [comment]
Could you check if you also have a crash like this? Is it improved after the update?
![Screenshot 2021-07-17 at 07 24 39](https://user-images.githubusercontent.com/2893584/126028090-99db288c-cc38-4bf4-8f2e-ceb2f4997026.png)

**Comment by phamsontm9x** [comment]
I don't have crash like this. My problem relate to dispatch_group_leave.

**Comment by nirlachman-monday** [comment]
+1 on this crash

**Comment by meberl-detego** [comment]
Any progress on this one with the newest versions?
Was it reproduce able with the latest versions as well?

Ok, just saw, that i can only use 3.8.0 as minimum version, since from 4.0.0 on, the minimum deployment target has been changed to 13.4.

**Comment by diemer** [comment]
I've found I'm able to reproduce this (after quite a bit of hoop jumping to get to the right place) if I start my app with wifi off and issue a number of log items. Looking to see if I can find a fix.

**Comment by diemer** [comment]
I think I've found a solution:
Change
```swift
task.write(logData, timeout: self.timeout) { [weak self] error in
    guard let self = self else {
        dispatchGroup.leave()
        return
    }

    self.dispatchQueue.async {
        if let error = error {
            sendStatus[tag] = error
        }

         dispatchGroup.leave()
    }
}
```

To
```swift
task.write(logData, timeout: self.timeout) { [weak self] error in
    guard let self = self else {
        dispatchGroup.leave()
        return
    }

    self.dispatchQueue.async(group: dispatchGroup) {
        if let error = error {
            sendStatus[tag] = error
        }
    }
}
```

Removing the explicit `dispatchGroup.leave()` in the `async` method, and instead passing the `dispatchGroup` in to the `async` method seems to fix things, at least in my testing. I'm guessing there are some edge cases in GCD that `async` passing in a group handles better than user code. 🤷‍♂️

Previously if the app has wifi off I can trigger this crash in some amount of time between 1 and 10 minutes.

After applying this change I've had a run of the app last 30 minutes with no crash before I manually restarted it to run a second test. I plan to run this second test overnight and if the outcome is positive I'll throw a PR up.

**Comment by albertodebortoli** [comment]
This has now been resolved with #125 and published in version 4.0.2.
